[SPIR-V][Sema] Requires nointerpolation on GetAttributeAtVertex param #0

Keenuts opened the pull request:
This commit adds new sema checks for SPIR-V to restrict the scope of the feature: the `nointerpolation` attribute is now required on all parameters used in GetAttributeAtVertex, and doesn't propagage magically to parent/child functions.

The current accepted cases this now blocks were mostly broken later in the codegen, so this should be an acceptable 'regression'.

Related to #6220
Fixes #6384, #6383, #6382

llvm-beanz commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -15176,6 +15176,83 @@ QualType Sema::getHLSLDefaultSpecialization(TemplateDecl *Decl) {
+    // to limit the scope, and fail gracefully in some cases.
+    if (!getLangOpts().SPIRV)
+      return false;
+    if (FDecl->getName() != "GetAttributeAtVertex")
```
Is this possible? I assume if the intrinsic op is set it should be this function.

Keenuts replied:
It just for a weird test in DXC, if an extension registers an instrinsic and there is a name conflict.

llvm-beanz replied:
I'm confused. Is there a test case where this fails? If we've assigned an intrinsic op then we're telling the compiler to treat it as that intrinsic regardless of the name of the function. I don't see anywhere else in DXC where we also separately need to verify the name of the function.

Keenuts replied:
IIRC the test was `ExtensionTest.EvalAttributeCollision`, which seems to register an intrinsic (collide_proc), with the same opcode as a registered intrinsic.
So here I hit the condition because the opcodes are the same, but the name mismatch.

llvm-beanz replied:
I see the test now. @adam-yang wrote it. I'm unsure what the context there is (although it terrifies me).

It shouldn't actually matter here though since that is a DXIL-only test and right above here you return if you're targeting DXIL, so this code shouldn't be possible to hit. We could make this an assert if you're concerned that it might come up for SPIR-V, but I think the inline SPIR-V proposal provides a more sane way to expose extensions than inserting a table of overridden intrinsics.

Keenuts replied:
Of course, forgot I gated this on `isSPIRV`. Cleaning that up now, adding an assert.

Keenuts commented:
Rebases on main to fix the bots.

hekota commented on tools/clang/test/SemaHLSL/getattribute-at-vertex.missing.nointerpolation.call.hlsl:
```diff
@@ -0,0 +1,11 @@
+// RUN: dxc -T ps_6_2 -E main %s -spirv -verify
```
The SemaHLSL tests need to run on both spirv and dxil backends. Could you please remove the -spirv argument?

hekota replied:
I see that this is a sema check specific to SPIR-V. In that case it would probably be best to have these tests in a subfolder with `lit.local.cfg` file that would have `config.unsupported = not config.spirv`.

llvm-beanz replied:
We can solve this by putting `// REQUIRES: spirv` in the test. That will make it only run if the spir-v code generator is enabled.

Keenuts replied:
There is one thing I don't get:
If we never add tests with `-spirv`, like in https://github.com/microsoft/DirectXShaderCompiler/pull/6467, we essentially lose coverage for the whole SPIR-V compilation side no?

As in, today, SEMA is shared, but if one day something goes bad, and for some reason sema diagnostics don't show up with `-spirv`, none of those tests would catch it.
Shouldn't all the sema tests also have a run line with `-spirv`?

llvm-beanz replied:
Counter point: If you look at clang tests for parsing and semantic analysis of C++ code don't have separate run lines for every target that LLVM supports.

Adding loads of duplicated test coverage that provides no additional meaningful test coverage does nothing but slow down development.

Keenuts replied:
I see, yes, that would be unscalable. Thanks!

Keenuts replied:
Added the `// REQUIRED: spirv` to the 2 impacted SEMA tests.

sudonatalie commented on tools/clang/test/CodeGenSPIRV/intrinsics.get-attribute-at-vertex.cbuf.var.hlsl:
```diff
@@ -0,0 +1,19 @@
+// RUN: dxc -T ps_6_2 -E main -spirv %s | FileCheck %s
```
Should be `RUN: %dxc`. Same applies for a few more tests below.

Keenuts replied:
Ah yes, fixed!

damyanp commented:
> Hello! Limited the scope of this change to the SPIR-V backend. Added some tests to make sure the DXIL side always catch those as part of the validation.

Could you update the PR description to reflect the change in scope please?

damyanp commented:
Also, can you confirm that with this change scoped to just affect SPIR-V, does it really address #6220?, or should we remove that from the description?

damyanp reviewed (changes requested):
Can you update the description to clarify if this impacts just SPIR-V or DXIL as well?

If it is just SPIR-V, can you double check the list of issues fixed by this, since one of them is a DXIL issue.

Keenuts commented on tools/clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -7732,6 +7732,8 @@ def warn_hlsl_implicit_vector_truncation  : Warning<
 def err_hlsl_nointerpolation_and_linear : Error<
     "nointerpolation cannot be used with any other interpolation mode specifier">;
+def err_hlsl_parameter_requires_attribute : Error<
+    "parameter %0 of %1 must have a a '%2' attribute">;
```
Ah nice catch.. Thanks!

Keenuts commented:
> Can you update the description to clarify if this impacts just SPIR-V or DXIL as well?
> 
> If it is just SPIR-V, can you double check the list of issues fixed by this, since one of them is a DXIL issue.

Right, forgot about the description.
Updated the issue list to only reference the one impacting both DXIL and SPIR-V, and changed title+description to say this only affects SPIR-V.
Thanks!

tex3d commented on tools/clang/test/SemaHLSL/getattribute-at-vertex.legal.hlsl:
```diff
@@ -0,0 +1,41 @@
+
+// expected-no-diagnostics
+struct S1 {
+  nointerpolation float3 f1;
```
I think there are missing cases where a struct field is marked nointerpolation and the field is passed to a function.  Then if you create a temporary of that structure in main and pass that field to a function.  What happens in these cases?

Keenuts replied:
Do you mean this case:

```hlsl
struct S6 {
  nointerpolation uint f1;
};

float4 main(S6 s6 : G) : SV_Target {
  S6 s6_copy = s6;
  float v8 = GetAttributeAtVertex(s6_copy.f1, 0);
  
  
  S6 s6_copy_2;
  s6_copy_2.f1 = s6.f1;
  float v8 = GetAttributeAtVertex(s6_copy_2.f1, 0);
}
```

If so, I can add those 2 cases. They pass the check as legal, but on the SPIR-V side, they crash (but that's another story 🙃)

tex3d replied:
You don't seem to have the common, legal case of passing `s6.f1` directly to a function that uses `GetAttributeAtVertex`, unless that was covered by some other existing test case I didn't see.  The temporary variable case was just an extension of this to see how it's handled.

Keenuts replied:
Right, added one line to check this.

tex3d reviewed (commented):
I have some concerns with the approach, but realize it is (likely too) late to voice them.  Since it's only enabled for SPIR-V, we have less grounds for objection, but we should be clear that this is not likely to be the solution for validating this scenario going forward.

`GetAttributeAtVertex` issues are in a class shared with `EvaluateAttribute*` intrinsics.  This change can't impact/help those intrinsics for those similar use cases.

At a high-level, I would prefer a warning diagnostic (default error on SPIR-V) that catches any use of this class of intrinsic outside the entry point (and reachable from an active entry point).  This would stop catch a lot more invalid use cases that we could not easily diagnose in SemaHLSL.  It would hit some cases that work today, but as a warning defaulting to error, it could be disabled for cases that are known to work for someone's specific scenario.  The idea is to put the warning up where we know the language can't handle this reliably today, even though it works in some cases today due to some unreliable/unpredictable "magic".

The language implications aren't consistent with HLSL's copy-in parameter passing, or the limited scope where interpolation modifiers have any meaning.  I think there are legitimate scenarios this would block, and other illegal ones that would be allowed through this check.  I think it would be good to at least make the diagnostic a warning, which defaults to error (and has a unique warning group).  This would allow someone to work around a blocked scenario when it doesn't impact `GetAttributeAtVertex`.

Also, how does this deal with implicitly `nointerpolation` inputs - any values of non-float types.  There's no requirement to use `nointerpolation` for those.

I can think of scenarios that might pass the diagnostic, and I don't know what the behavior will be for these.  I think we could use a few more some test cases for this.

damyanp reviewed (approved):
Thanks for updating the description.  It looks like Tex has also added some notes on the code.

Keenuts commented on tools/clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -5288,7 +5288,10 @@ Sema::BuildResolvedCallExpr(Expr *Fn, NamedDecl *NDecl,
     if (CheckFunctionCall(FDecl, TheCall, Proto))
       return ExprError();
-
+    if (CheckHLSLFunctionCall(FDecl, TheCall))
+      return ExprError();
+    if (hlsl::IsIntrinsicOp(FDecl) && CheckHLSLIntrinsicCall(FDecl, TheCall))
+      return ExprError();
```
What do you mean? Is it called below it, only gated by the `IsInstrinsicOp` call.
Am I missing something?

tex3d replied:
I meant that we could have one hook function into SemaHLSL (CheckHLSLFunctionCall), and the knowledge of using IsIntrinsicOp and CheckHLSLIntrinsicCall would be localized to CheckHLSLFunctionCall in SemaHLSL, instead of requiring this to be known in SemaExpr.  Also, CheckHLSLIntrinsicCall wouldn't have to be part of Sema.  That is unless you can think of a place that CheckHLSLIntrinsicCall would need to be called from Sema without calling CheckHLSLFunctionCall.

Keenuts replied:
Ah right, now I understand the 'under' 😊 Moved the call inside the CheckHLSLFunctionCall!

Keenuts commented:
> At a high-level, I would prefer a warning diagnostic (default error on SPIR-V) that catches any use of this class of intrinsic outside the entry point (and reachable from an active entry point). This would stop catch a lot more invalid use cases that we could not easily diagnose in SemaHLSL.

I'd be fine with blocking that too.

> I think there are legitimate scenarios this would block, and other illegal ones that would be allowed through this check. 

My point of view on that is: the feature cannot be implemented. It is ill-defined to have something clean.
We tried to implement it still, but we have many edge cases we fail to handle, and as a result, the compiler either crashes, or errors with a validation error (which on the the SPIR-V side, is considered a compiler bug).

If I'm building an engine, and use DXC, I'd rather see a feature be disabled with a clear error message, than having to deal with validation errors, compiler instabilities, or guess in which scenario can this feature be used.

But I understand removing the feature is a bit extreme, so what I'm trying is to at least drastically reduce the scope of the feature, limiting the magic required to make it work.

Only allowing the function in the entrypoint would for example go in that direction: remove most of the ill-defined scenarios, to have something predictable and usable.

Keenuts commented on tools/clang/lib/SPIRV/PervertexInputVisitor.cpp:
```diff
@@ -56,6 +56,11 @@ int PervertexInputVisitor::appendIndexZeroAt(
+  // If there is no AST type, it means the type was already lowered from some
+  // expression/construct. If it required expansion, it should be done already.
+  if (!spvInst->hasAstResultType())
+    return spvInst->isNoninterpolated();
```
Rebased on main to clean this up.

sudonatalie commented:
> My point of view is: the feature is ill-defined, and cannot be implemented as is. Knowing that, we tried to implement it. As a result, we have many edge cases we fail to handle: the compiler either crashes, or displays a validation error (which on the the SPIR-V side, is considered a compiler bug).

+1. We've spent a lot of time trying to figure out a reasonable way to match the DXIL behavior in SPIR-V and it's just not feasible with the current semantics. This PR will get the SPIR-V backend to a better state of erroring instead of crashing or throwing unparseable validation errors. If we get to a re-design in https://github.com/microsoft/hlsl-specs/issues/181 then we'll have a good foundation to implement the SPIR-V backend from but I think this is the best option in the meantime.
